src/xrpld/rpc/handlers/DepositAuthorized.cpp

```diff
@@ -88,23 +90,98 @@ doDepositAuthorized(RPC::JsonContext& context)
+        auto const& creds(params[jss::credentials]);
+        if (!creds.isArray() || !creds ||
+            (creds.size() > maxCredentialsArraySize))
+        {
+            return RPC::make_error(
+                rpcINVALID_PARAMS,
+                RPC::expected_field_message(
+                    jss::credentials, "an array of CredentialID(hash256)"));
+        }
+
+        for (auto const& jo : creds)
         {
-            // See if a preauthorization entry is in the ledger.
-            auto const sleDepositAuth =
-                ledger->read(keylet::depositPreauth(dstAcct, srcAcct));
-            depositAuthorized = static_cast<bool>(sleDepositAuth);
+            if (!jo.isString())
+            {
+                return RPC::make_error(
+                    rpcINVALID_PARAMS,
+                    RPC::expected_field_message(
+                        jss::credentials, "an array of CredentialID(hash256)"));
+            }
+
+            uint256 credH;
+            auto const credS = jo.asString();
+            if (!credH.parseHex(credS))
+            {
+                return RPC::make_error(
+                    rpcINVALID_PARAMS,
+                    RPC::expected_field_message(
+                        jss::credentials, "an array of CredentialID(hash256)"));
+            }
+
+            std::shared_ptr<SLE const> sleCred =
+                ledger->read(keylet::credential(credH));
+            if (!sleCred || (sleCred->getType() != ltCREDENTIAL) ||
+                !(sleCred->getFlags() & lsfAccepted))
+            {
+                RPC::inject_error(rpcBAD_CREDENTIALS, result);
+                return result;
+            }
+
+            if (credentials::checkExpired(
+                    sleCred, context.app.timeKeeper().now()))
+            {
+                RPC::inject_error(rpcBAD_CREDENTIALS, result);
+                return result;
+            }
+
+            if (reqAuth)
+            {
+                auto credential = STObject::makeInnerObject(sfCredential);
+                credential.setAccountID(sfIssuer, (*sleCred)[sfIssuer]);
+                credential.setFieldVL(
+                    sfCredentialType, (*sleCred)[sfCredentialType]);
+                authCreds.push_back(std::move(credential));
+            }
         }
     }
+
+    // If the two accounts are the same OR if that flag is
+    // not set, then the deposit should be fine.
+    bool depositAuthorized = true;
+
+    if (reqAuth)
+    {
+        if (credentialsPresent)
+        {
+            auto const sorted = credentials::makeSorted(authCreds);
+            if (sorted.empty())
+            {
+                RPC::inject_error(rpcBAD_CREDENTIALS, result);
+                return result;
+            }
+
+            depositAuthorized =
+                ledger->exists(keylet::depositPreauth(dstAcct, sorted));
+        }
+        else
+            depositAuthorized =
+                ledger->exists(keylet::depositPreauth(dstAcct, srcAcct));
```

Comment:
IMO this RPC should check for both `DepositPreauth` objects (credentials and non-credentials) if credentials are provided.

Reply:
It will conflict with internal behavior that provided credentials doesn't count only if account doesn't require authorization (no flag set, or src==dst).  If credentials provided and there is only  "old way" DepositPreauth it will fail.

Reply:
Behavior changed to check for both DepositPreauth objects -  old one and new